src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzersOptions.cs

```diff
@@ -48,6 +49,14 @@ public sealed class CompilationWithAnalyzersOptions
         /// </summary>
         public bool ReportSuppressedDiagnostics => _reportSuppressedDiagnostics;
 
+        /// <summary>
+        /// Callback to allow individual analyzers to have their own <see cref="AnalyzerOptions"/> distinct
+        /// from the shared instance provided in <see cref="Options"/>.  If <see langword="null"/> then <see cref="Options"/>
+        /// will be used for all analyzers.
+        /// </summary>
+        public Func<DiagnosticAnalyzer, AnalyzerOptions>? AnalyzerSpecificOptionsFactory
```

Comment:
One thing I noticed while prototyping this myself was AnalyzerOptions holds the list of additional documents, so allowing the _entire_ options to replaced could have potentially strange behaviors if the set of additional documents were to change. I ended up making this a `Func<DiagnosticAnalyzer, AnalyzerConfigOptionsProvider>` so that way _only_ the options are changeable.

Reply:
had the same thought.  it is weird to have the .AdditionalFiles.  i'll take your approach and just do the AnalyzerConfigOptionsProvider approach.